source/PluginHelper.cpp

```diff
@@ -0,0 +1,220 @@
+#else
+			FILE *out = fopen(location.c_str(), "wb");
+#endif
+			// Set the url that gets downloaded
+			curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+			// Follow redirects
+			curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1l);
+			// How long we will wait
+			curl_easy_setopt(curl, CURLOPT_CA_CACHE_TIMEOUT, 604800L);
+			// Set the write function and the output file used in the write function
+			curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteData);
+			curl_easy_setopt(curl, CURLOPT_WRITEDATA, out);
+
+			res = curl_easy_perform(curl);
+			if(res != CURLE_OK)
+				fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
+			curl_easy_cleanup(curl);
+
+			fclose(out);
+		}
+
+		curl_global_cleanup();
+
+		return res == CURLE_OK;
+	}
+
+
+
+	// Copy an entry from one archive to the other
+	int CopyData(struct archive *ar, struct archive *aw)
+	{
+		int retVal;
+		const void *buff;
+		size_t size;
+#if ARCHIVE_VERSION_NUMBER >= 3000000
+		int64_t offset;
+#else
+		off_t offset;
+#endif
+
+		for(;;)
+		{
+			retVal = archive_read_data_block(ar, &buff, &size, &offset);
+			if(retVal == ARCHIVE_EOF)
+				return (ARCHIVE_OK);
+			if(retVal != ARCHIVE_OK)
+				return (retVal);
+			retVal = archive_write_data_block(aw, buff, size, offset);
+			if(retVal != ARCHIVE_OK)
+			{
+				printf("archive_write_data_block(), %s", archive_error_string(aw));
+				return (retVal);
+			}
+		}
+	}
+
+
+
+	bool ExtractZIP(string filename, string destination, string expectedName)
+	{
+		struct archive *archive;
+		struct archive *ext;
+		struct archive_entry *entry;
+		int retVal;
+
+		int flags;
+		flags = ARCHIVE_EXTRACT_TIME;
+		flags |= ARCHIVE_EXTRACT_PERM;
+		flags |= ARCHIVE_EXTRACT_ACL;
+		flags |= ARCHIVE_EXTRACT_FFLAGS;
+
+		// Create the handles for reading/writing
+		archive = archive_read_new();
+		ext = archive_write_disk_new();
+		archive_write_disk_set_options(ext, flags);
+		archive_read_support_format_all(archive);
+		if(!filename.empty() && strcmp(filename.c_str(), "-") == 0)
+			filename = "";
+		if((retVal = archive_read_open_filename(archive, filename.c_str(), 10240)))
+		{
+			printf("archive_read_open_filename(), %s, %i", archive_error_string(archive), retVal);
+			return false;
+		}
+
+		// Check if this plugin has the right head folder name
+		retVal = archive_read_next_header(archive, &entry);
+		string firstEntry = archive_entry_pathname(entry);
+		bool fitsExpected = firstEntry == (expectedName);
+		archive_read_data_skip(archive);
+		// Check if this plugin has a head folder, if not create one in the destination
+		retVal = archive_read_next_header(archive, &entry);
+		string secondEntry = archive_entry_pathname(entry);
+		bool hasHeadFolder = secondEntry.find(firstEntry) != std::string::npos;
+		if(!hasHeadFolder)
+#if defined(_WIN32)
+			_wmkdir(Utf8::ToUTF16(destination + expectedName).c_str());
+#else
+			mkdir((destination + expectedName).c_str(), 0777);
```

Comment:
Isn't libarchive going to create the directory anyways if it doesn't exist?

Reply:
not too sure about that